pandas/_libs/groupby.pyx

```diff
@@ -389,6 +390,8 @@ def group_fillna_indexer(ndarray[int64_t] out, ndarray[int64_t] labels,
             else:  # reset items when not missing
                 filled_vals = 0
                 curr_fill_idx = idx
+            if dropna and labels[idx] == -1:
+                curr_fill_idx = -1
```

Comment:
I believe you can move these two lines to the top of the if-else, and change the `if mask[idx] == 1:` to `elif mask[idx] == 1`. This way we're not executing code needlessly.

Reply:
Done